### PR TITLE
Get avr-gcc@11 to work on Apple ARM systems.

### PR DESCRIPTION
**Pull request by DavidEGrayson**
I have a Mac Mini with the new M1 chip and I was able to get the avr-gcc@11.rb formula to build and function properly by making the changes here.

This adds a patch that fixes the missing definition of host_hooks in GCC.  I added that patch to the source tree of this project because that seems like the right place to put it, but I couldn't easily figure out how to fetch the patch from there, so I'm just fetching it from the Internet (like all the other patches used in this project).

For avr-libc, there is no need to provide the `--build` argument.  The avr-libc configure script is perfectly capable of running its own `config.guess` and `config.sub` scripts to figure out what system it is being built on.  So I removed that `--build` argument.  However, avr-libc ships with old versions of those scripts which give invalid results on Apple ARM systems.  Specifically, `config.guess` returns `-apple-darwin20.1.0` and then `config.sub` has an error because that's an invalid system name.  We could download the latest versions of those scripts from GNU and put them into avr-libc, but that sounded tricky, so instead I just run config.guess, see if the problem is going to happen, and then set an environment variable that bypasses both of those scripts if needed.

I used the `file` utility to check that my compiled avr-gcc binary is actually an ARM binary, and removed the caveat about possibly needing Rosetta 2.

Someone should probably test my changes on an Intel Mac using a procedure like this:

1. `brew install --formula ./Formula/avr-gcc@11.rb`
2. Follow the instructions that get printed to add avr-gcc 11 to your PATH.
3. Run `which avr-gcc` to make sure it is the one you just built.
4. Run `brew test avr-gcc@11`.

### CODE REVIEW

**Comment by ladislas** [comment]
Thanks for the PR @DavidEGrayson!

The CI fails because of style issues.

Can you run:

```bash
# check style issues
brew style osx-cross/avr

# fix style issues
brew style --fix osx-cross/avr

# check again and fix manually if needed
brew style osx-cross/avr

# audit formulae
brew audit --tap=osx-cross/avr
```

Then push with the changes so I can run CI again.

Also no need to provide the patch, just a comment about the origin (gcc.rb from brew/core I guess) would be enough. You can then remove the patch file.

Thanks a lot!

**Comment by DavidEGrayson** [comment]
OK, I fixed the style issues.  I also got the tests to work by providing the full path to `avr-objcopy` (and wrapped that line to avoid a style issue).

File: Formula/avr-gcc@11.rb [hunk 1]
```diff
@@ -63,6 +61,14 @@ class AvrGccAT11 < Formula
     end
   end
 
+  # This patch fixes a GCC compilation error on Apple ARM systems by adding
+  # a defintion for host_hooks.  Patch comes from
+  # https://github.com/riscv/riscv-gnu-toolchain/issues/800#issuecomment-808722775
+  patch do
+    url "https://gist.githubusercontent.com/DavidEGrayson/88bceb3f4e62f45725ecbb9248366300/raw/c1f515475aff1e1e3985569d9b715edb0f317648/gcc-11-arm-darwin.patch"
+    sha256 "c4e9df9802772ddecb71aa675bb9403ad34c085d1359cb0e45b308ab6db551c6"
+  end
```
Comment (ladislas) [inline Formula/avr-gcc@11.rb:70]: So you're sure that without this part, it doesn't work/compile on M1?

File: Formula/avr-gcc@11.rb [hunk 2]
```diff
@@ -132,22 +138,20 @@ def install
       ENV.delete "CC"
       ENV.delete "CXX"
 
+      # avr-libc ships with outdated config.guess and config.sub scripts that
+      # do not support Apple ARM systems, causing the configure script to fail.
       build_config = `./config.guess`.chomp
+      if build_config.start_with?("-")
```
Comment (ladislas) [inline Formula/avr-gcc@11.rb]: why `"-"`?

File: Formula/avr-gcc@11.rb [hunk 3]
```diff
@@ -177,7 +181,8 @@ def caveats
 
     system "#{bin}/avr-gcc", "-mmcu=atmega328p", "-Os", "-c", "hello.c", "-o", "hello.c.o", "--verbose"
     system "#{bin}/avr-gcc", "hello.c.o", "-o", "hello.c.elf"
-    system "avr-objcopy", "-O", "ihex", "-j", ".text", "-j", ".data", "hello.c.elf", "hello.c.hex"
+    system "#{Formula["avr-binutils"].opt_bin}/avr-objcopy", "-O", "ihex", "-j", ".text", "-j", ".data", \
```
Comment (ladislas) [inline Formula/avr-gcc@11.rb:183]: why is this needed?

File: Formula/avr-gcc@11.rb [hunk 4]
```diff
@@ -219,7 +224,8 @@ def caveats
 
     system "#{bin}/avr-g++", "-mmcu=atmega328p", "-Os", "-c", "hello.cpp", "-o", "hello.cpp.o", "--verbose"
     system "#{bin}/avr-g++", "hello.cpp.o", "-o", "hello.cpp.elf"
-    system "avr-objcopy", "-O", "ihex", "-j", ".text", "-j", ".data", "hello.cpp.elf", "hello.cpp.hex"
+    system "#{Formula["avr-binutils"].opt_bin}/avr-objcopy", "-O", "ihex", "-j", ".text", "-j", ".data", \
```
Comment (ladislas) [inline Formula/avr-gcc@11.rb:226]: same as above?

**Review by ladislas — COMMENTED** [review]
Thanks a lot for the fixes! Overall it looks good, I still have a few questions about the changes though. 

Were you able to try it on both Intel and M1 Macs with a real program? Does it work?

**Comment by ladislas** [comment]
cc @osx-cross/team @brucehoult @braindaddy73 @Grippy98 @jknoetzke

could you guys take a look and give this a try?

related #220 #222 #232

File: Formula/avr-gcc@11.rb (hunk 1 above)
Comment (DavidEGrayson) [reply Formula/avr-gcc@11.rb:70]: Yes, I was getting an error about `_host_hooks` not being defined when I tried to build on an M1 machine.

- [Here's the error I saw.](https://github.com/osx-cross/homebrew-avr/issues/237#issuecomment-859816991)
- [Here are other people getting the same error.](https://github.com/riscv/riscv-gnu-toolchain/issues/800#issuecomment-808722775)

File: Formula/avr-gcc@11.rb (hunk 2 above)
Comment (DavidEGrayson) [reply Formula/avr-gcc@11.rb]: I saw that config.guess returned `-apple-darwin20.1.0` and then config.sub had an error because that's an invalid [triplet](https://wiki.osdev.org/Target_Triplet).   Of course I can't hardcode that string but I wanted to come up with some simple way to detect the problem.

Any triplet that starts with `-` is probably wrong, because the triplet should start with the name of the machine that is being targeted.  So it seemed fine to just use that as the condition.  

I couldn't figure out how to ask Homebrew whether we are building on an M1 machine.

File: Formula/avr-gcc@11.rb (hunk 3 above)
Comment (DavidEGrayson) [reply Formula/avr-gcc@11.rb:183]: I wanted to run the tests that come with this formula in order to have more confidence in my changes, but when I did that, Homebrew printed an error message to the effect that it was unable to run `avr-objcopy`.  I was able to fix it by making this change.  I'm pretty sure that avr-objcopy was not found because it was not on my PATH.  When these tests run, the PATH environment variable is unset, because `ENV.clear` is called at the top of the tests.

(I think the only reason these tests ever worked for anyone was that the shell or some other low-level part of the system probably added some default folders to the PATH, allowing the test script to find the two executables on the PATH that it needs: cat and avr-objcopy.  But on my machine, I didn't have `avr-objcopy` on the PATH.)

I considered putting the AVR binutils onto the PATH at the top of the test.  However, I'm not sure if the AVR binutils will always be on the user's PATH when they are using this toolchain, so I didn't want to make the test environment be like that.

Comment (DavidEGrayson) [reply Formula/avr-gcc@11.rb:183]: Apparently writing `ENV.clear` is not sufficient to prevent the test script from finding various executables on your system like `cat` and `avr-objcopy` and using them.  If you want the tests to run in a more predictable way, the PATH should probably be set to a string that contains a very obscure folder name that will never exist.

File: Formula/avr-gcc@11.rb (hunk 4 above)
Comment (DavidEGrayson) [reply Formula/avr-gcc@11.rb:226]: Yes.

File: Formula/avr-gcc@11.rb (hunk 3 above)
Comment (DavidEGrayson) [reply Formula/avr-gcc@11.rb:183]: I used `opt_bin` because that is how we find the AVR binutils in another part of this formula.  I was unable to find information on the differences between `bin` and `opt_bin` and both seem to work fine.

**Comment by DavidEGrayson** [comment]
> Were you able to try it on both Intel and M1 Macs with a real program? Does it work?

I was only able to try it on an M1 Mac.  I tested it before submitting my original pull request, and I was able to compile a program that runs successfully on a [Pololu robot controller](https://www.pololu.com/product/1227) and controls an LCD.

I think the biggest risk was that adding `host_hooks` might cause a multiple definition error when building GCC.  But if your CI was running on an Intel Mac and that didn't happen, then it's probably fine.

File: Formula/avr-gcc@11.rb (hunk 1 above)
Comment (ladislas) [reply Formula/avr-gcc@11.rb:70]: thanks!

File: Formula/avr-gcc@11.rb (hunk 2 above)
Comment (ladislas) [reply Formula/avr-gcc@11.rb]: You can use `Hardware::CPU.arm?`, it would be better than the triplet check.

See https://github.com/Homebrew/homebrew-core/blob/master/Formula/gcc.rb#L74

File: Formula/avr-gcc@11.rb (hunk 3 above)
Comment (ladislas) [reply Formula/avr-gcc@11.rb:183]: > I was unable to find information on the differences between bin and opt_bin and both seem to work fine.

- `opt_bin` --> `/usr/local/opt/foo/bin`
- `bin` --> `/usr/local/Cellar/foo/0.1/bin`

Okay, now I understand the why. But I'm not sure it's needed for everyone, as in your case I'm surprised that there is no avr-objcopy in your $PATH.

avr-binutils must be installed with avr-gcc otherwise the toolchain is completely useless. avr-binutils can be shared by different version of avr-gcc, this is why it's only pulled once the first time.

avr-gcc automatically pulls avr-binutils with the `depends_on "avr-binutils"`, so avr-binutils must be in the path.

This is the normal process:

Screenshot 2021-06-16 at 14 39 36

And this is the test:

Screenshot 2021-06-16 at 14 40 50

If you `brew remove` everything osx-cross/avr related and start fresh, what happens?

Comment (DavidEGrayson) [reply Formula/avr-gcc@11.rb:183]: > avr-binutils must be installed with avr-gcc otherwise the toolchain is completely useless.

True, but it does not have to be on the PATH.  GCC is configured with the
`--with-as` and `--with-ld` options to tell it where to find those utilities.
With those two utilities, avr-gcc is able to create an ELF file, and you can
actually use AVRDUDE directly with ELF files instead of converting it to HEX.

OK, I did a new test today.  I changed the test in the avr-gcc@11.rb formula
on my computer back to the way it was, so it does not provide the full path
to avr-objcopy during the test.  I have attached a shell session showing:

1. I uninstalled all the software from this repo.
2. I was able to install the avr-gcc@11.rb formula.
3. The toolchain itself works, producing a HEX file that is the
   same as `hello_c_hex` from the test if I manually compile the
   source code from the test.
4. `/opt/homebrew/bin/avr-objcopy` is actually on the PATH in my shell.
5. The tests still fail when running `avr-objcopy`
   (presumably because the PATH gets unset at the top of the test).
   The error message is pretty vague, but you can see it in the attached
   shell session.

I have to use slightly different commands than you because I'm working with a
formula on my computer instead of on GitHub.

Attachment: [shell_session.txt](https://github.com/osx-cross/homebrew-avr/files/6664614/shell_session.txt)

**Comment by DavidEGrayson** [comment]
In my latest commit, I changed the code so it does not run `config.guess` but instead just asks Homebrew if we are compiling for ARM.  

```ruby
      if Hardware::CPU.arm?
         ENV["ac_cv_build"] = "aarch64-apple-darwin"
         puts "Forcing build system to aarch64-apple-darwin."
```

I changed the triple to `aarch64-apple-darwin` because that is what I get when I run the latest `config.guess` script from GNU on my M1 Mac Mini.  If we care about being able to compile this formula on ARM Linux platforms, maybe we should also find some way to ensure the operating system is actually Darwin, but I'm not sure if that's important and I suspect avr-libc would still build fine anyway.

**Comment by DavidEGrayson** [comment]
In my latest commit (which I will push very soon), I changed the code so it does not run `config.guess` but instead just asks Homebrew if we are compiling for an ARM macOS machine.  

```ruby
      if OS.mac? && Hardware::CPU.arm?
        ENV["ac_cv_build"] = "aarch64-apple-darwin"
        puts "Forcing build system to aarch64-apple-darwin."
      end
```

I changed the triple to `aarch64-apple-darwin` because when I run the latest `config.guess` script from GNU on my M1 Mac Mini, it returns `aarch64-apple-darwin20.1.0`.  We can remove this workaround once avr-libc updates their copy of `config.guess`.

File: Formula/avr-gcc@11.rb (hunk 3 above)
Comment (ladislas) [reply Formula/avr-gcc@11.rb:183]: > True, but it does not have to be on the PATH.  GCC is configured with the
> `--with-as` and `--with-ld` options to tell it where to find those utilities.

Ah true! That settles it then!, let's keep the `opt_bin` thing for consistency.

**Comment by ladislas** [comment]
Thanks for the explanations, I'll run CI again. When it's green, can you squash/fixup the last two commits into one?

**Comment by DavidEGrayson** [comment]
OK.  I combined the last two commits, so there are only two commits total in this pull request now.

**Comment by ladislas** [comment]
Thanks again for the good work! Just merged it! :)

**Comment by DavidEGrayson** [comment]
Thank you for getting this merged in so quickly!  Now I'm working on a pull request to do the same things to the over versions of GCC.
